weave/flow/eval_imperative.py

```diff
@@ -0,0 +1,314 @@
+    # Check if name is not empty
+    if not name:
+        raise ValueError("Scorer name cannot be empty")
+
+    # Check if name follows Python class naming conventions
+    # Class names should start with a letter or underscore and contain only
+    # alphanumeric characters and underscores
+    import re
+
+    if not re.match(r"^[a-zA-Z_][a-zA-Z0-9_]*$", name):
+        raise ValueError(
+            f"Invalid scorer name: '{name}'. Scorer names must start with a letter or underscore "
+            "and contain only alphanumeric characters and underscores."
+        )
+
+    # Check if name is not a Python keyword
+    import keyword
+
+    if keyword.iskeyword(name):
+        raise ValueError(f"Scorer name '{name}' cannot be a Python keyword")
+
+    return name
+
+
+class ImperativeScoreLogger(BaseModel):
+    """This class provides an imperative interface for logging scores."""
+
+    # model_id: ID
+    inputs: dict
+    output: Any
+    predict_and_score_call: Call
+    evaluate_call: Call
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def log_score(self, scorer: Scorer | str, score: ScoreType) -> None:
+        """Log a score synchronously by calling the async method.
+
+        This is a convenience method for when you don't want to use async/await.
+        If called from within an existing event loop, use _alog_score instead.
+        """
+        try:
+            asyncio.run(self._alog_score(scorer, score))
+        except RuntimeError as e:
+            if "This event loop is already running" in str(e):
+                raise RuntimeError(
+                    "Cannot call log_score from an async context. Use _alog_score instead."
+                ) from e
+            raise
+
+    @validate_call
+    async def _alog_score(
+        self,
+        scorer: Annotated[
+            Scorer | dict | str,
+            BeforeValidator(_cast_to_cls(Scorer)),
+            Field(
+                description="A metadata-only scorer used for comparisons."
+                "Alternatively, you can pass a dict of attributes or just a string"
+                "representing the ID of your scorer."
+            ),
+        ],
+        score: ScoreType,
+    ) -> None:
+        # this is safe; pydantic casting is done in validator above
+        scorer = cast(Scorer, scorer)
+
+        @weave.op(name=scorer.name)
+        def score_method(self: Scorer, *, output: Any, **inputs: Any) -> ScoreType:
+            # TODO: can't use score here because it will cause version mismatch
+            # return score
+            return cast(ScoreType, current_score.get())
+
+        scorer.__dict__["score"] = MethodType(score_method, scorer)
+
+        # attach the score feedback to the predict call
+        with call_context.set_call_stack(
+            [self.evaluate_call, self.predict_and_score_call]
+        ):
+            with _set_current_score(score):
+                await self.predict_and_score_call.apply_scorer(scorer)
```

Comment:
We don't apply scorer to the predict and score - we apply it to the predict call